File: template/kmp.java
public int[] pre(String pattern){
        int pre[]=new int[pattern.length()];
        int l=0,r=1;
        while(r<pattern.length()){
            if(pattern.charAt(r)==pattern.charAt(l)){
                pre[r]=l+1;
                l++;
                r++;
            }else{
                if(l==0)r++;
                else{
                    l=pre[l-1];
                }
            }
        }
        return pre;
    }

File: template/kmp.cpp
vector<int> pre(string& pattern){
  vector<int> pre(pattern.size());
  int l=0,r=1;
  while(r < pattern.size()()){
    if(pattern[r] == pattern[l]){
      pre[r] = l + 1;
      l++;
      r++;
    }else{
      if(l == 0)r++;
      else{
            l = pre[l-1];
          }
    }
  }
  return pre;
}
